rust/arrow/src/ipc/reader.rs

```
@@ -521,7 +524,111 @@ impl<R: Read + Seek> Reader<R> {
+        let mut reader = BufReader::new(reader);
+        // determine metadata length
+        let mut meta_size: [u8; 4] = [0; 4];
+        reader.read_exact(&mut meta_size)?;
+        let meta_len = u32::from_le_bytes(meta_size);
+
+        let mut meta_buffer = vec![0; meta_len as usize];
+        reader.read_exact(&mut meta_buffer)?;
+
+        let vecs = &meta_buffer.to_vec();
+        let message = ipc::get_root_as_message(vecs);
+        // message header is a Schema, so read it
+        let ipc_schema: ipc::Schema = message
+            .header_as_schema()
+            .expect("Arrow Stream must start with a schema");
+        let schema = ipc::convert::fb_to_schema(ipc_schema);
+
+        Ok(Self {
+            reader,
+            schema: Arc::new(schema),
+            finished: false,
+        })
+    }
+
+    /// Return the schema of the stream
+    pub fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+
+    /// Read the next record batch
+    pub fn next(&mut self) -> Result<Option<RecordBatch>> {
+        if self.finished {
+            return Ok(None);
+        }
+        // determine metadata length
+        let mut meta_size: [u8; 4] = [0; 4];
+        self.reader.read_exact(&mut meta_size)?;
+        let meta_len = u32::from_le_bytes(meta_size);
+
+        if meta_len == 0 {
+            // the stream has ended, mark the reader as finished
+            self.finished = true;
+            return Ok(None);
+        }
+
+        let mut meta_buffer = vec![0; meta_len as usize];
+        self.reader.read_exact(&mut meta_buffer)?;
+
+        let vecs = &meta_buffer.to_vec();
+        let message = ipc::get_root_as_message(vecs);
+
+        match message.header_type() {
+            ipc::MessageHeader::Schema => {
+                panic!("Not expecting a schema when messages are read")
+            }
+            ipc::MessageHeader::RecordBatch => {
+                let batch = message.header_as_record_batch().unwrap();
```

We should propagate this error instead of calling `unwrap`.

---

Done